### PR TITLE
[2201.3.x] Fix compiler-plugin logic to support recursive payload types with unions

### PR DESCRIPTION
**Pull request by ayeshLK**
Fixes: [#3966](https://github.com/ballerina-platform/ballerina-standard-library/issues/3966)

## Examples
N/A

## Checklist
- [X] Linked to an issue
- [X] Updated the changelog
- [x] Added tests
- [ ] ~Updated the spec~
- [ ] ~Checked native-image compatibility~

### CODE REVIEW

File: ballerina-tests/Dependencies.toml [hunk 1]
```diff
@@ -38,7 +38,7 @@ dependencies = [
 [[package]]
 org = "ballerina"
 name = "constraint"
-version = "1.0.1"
+version = "1.0.2"
```
Comment (ayeshLK) [inline ballerina-tests/Dependencies.toml:41]:
```suggestion
version = "1.0.1"
```

File: ballerina-tests/Dependencies.toml [hunk 2]
```diff
@@ -309,7 +309,7 @@ modules = [
 [[package]]
 org = "ballerina"
 name = "log"
-version = "2.5.0"
+version = "2.5.1"
```
Comment (ayeshLK) [inline ballerina-tests/Dependencies.toml:312]:
```suggestion
version = "2.5.0"
```

File: ballerina/Dependencies.toml [hunk 3]
```diff
@@ -39,7 +39,7 @@ modules = [
 [[package]]
 org = "ballerina"
 name = "constraint"
-version = "1.0.1"
+version = "1.0.2"
```
Comment (ayeshLK) [inline ballerina/Dependencies.toml:42]:
```suggestion
version = "1.0.1"
```

File: ballerina/Dependencies.toml [hunk 4]
```diff
@@ -237,7 +237,7 @@ modules = [
 [[package]]
 org = "ballerina"
 name = "log"
-version = "2.5.0"
+version = "2.5.1"
```
Comment (ayeshLK) [inline ballerina/Dependencies.toml:240]:
```suggestion
version = "2.5.0"
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpResourceValidator.java [hunk 5]
```diff
@@ -656,10 +656,21 @@ private static boolean isValidRecordParam(List<TypeSymbol> parentTypes, TypeSymb
...
             TypeSymbol effectiveTypeDescriptor = ((TypeReferenceTypeSymbol) typeDescriptor).typeDescriptor();
+            List<TypeSymbol> newParentTypes = getParentTypes(effectiveTypeDescriptor, parentTypes);
             if (effectiveTypeDescriptor.typeKind() == TypeDescKind.RECORD) {
-                List<TypeSymbol> newParentTypes = getParentTypes(effectiveTypeDescriptor, parentTypes);
                 return validateRecordTypeField(
                         newParentTypes, ctx, paramLocation, paramName, (RecordTypeSymbol) effectiveTypeDescriptor);
+            } else if (effectiveTypeDescriptor.typeKind() == TypeDescKind.UNION) {
+                List<TypeSymbol> typeDescriptors = ((UnionTypeSymbol) effectiveTypeDescriptor).memberTypeDescriptors();
+                return typeDescriptors.stream()
+                        .allMatch(td -> isValidRecordParam(newParentTypes, td, ctx, paramLocation, paramName));
```
Comment (chamil321) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpResourceValidator.java:666]:
```suggestion
                        .allMatch(typeDescriptor -> isValidRecordParam(newParentTypes, typeDescriptor, ctx, paramLocation, paramName));
```

**Review by chamil321 — APPROVED** [review]

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpResourceValidator.java (hunk 5 above)
Comment (ayeshLK) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpResourceValidator.java]: Fixed with [71233e938d67b43bb0eaa324a31d69787c3fd473](https://github.com/ballerina-platform/module-ballerina-http/pull/1422/commits/71233e938d67b43bb0eaa324a31d69787c3fd473)

**Review by chamil321 — APPROVED** [review]

**Pull request merged**